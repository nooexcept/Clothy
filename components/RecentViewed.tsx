import React from 'react'
import Link from 'next/link'
import Card from '@material-ui/core/Card'
import CardActionArea from '@material-ui/core/CardActionArea'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import { Product } from '../src/types'

interface RecentViewedProps {
    recentProducts: Array<Product>
}

const RecentViewed: React.SFC<RecentViewedProps> = ({ recentProducts }) => {
    return (
        <>
            {recentProducts.map((prod: Product) => {
                return (
                    <Grid item xs={3} key={prod.url}>
                        <Link href={`products/${prod.url}`}>
                            <Card>
                                <CardActionArea>
                                    <CardMedia
                                        component="img"
                                        alt={prod.name}
                                        image={prod.images[0]}
                                        title={prod.name}
                                    />
                                    <CardContent>
                                        <Typography
                                            gutterBottom
                                            variant="h5"
                                            component="h2"
                                        >
                                            {prod.name}
                                        </Typography>
                                    </CardContent>
                                </CardActionArea>
                            </Card>
                        </Link>
                    </Grid>
                )
            })}
        </>
    )
}

export default RecentViewed
