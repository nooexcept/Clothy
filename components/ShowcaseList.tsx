import * as React from 'react'
import { makeStyles, createStyles } from '@material-ui/core/styles'
import Grid from '@material-ui/core/Grid'
import Zoom from '@material-ui/core/Zoom'

const useStyles = makeStyles(() =>
    createStyles({
        root: {
            flexGrow: 1,
            maxWidth: '100%',
        },
    })
)

interface ShowcaseListProps {
    showcases: Array<React.ReactElement>
}

export default function ShowcaseList(
    props: ShowcaseListProps
): React.ReactElement {
    const classes = useStyles()

    return (
        <Grid container className={classes.root} spacing={2}>
            <Grid item xs={12}>
                <Grid container justify="center" spacing={2}>
                    {props.showcases.map((showcase, i) => (
                        <Zoom
                            in
                            key={i}
                            style={{
                                transitionDelay: `${100 + 100 * i}ms`,
                            }}
                        >
                            <Grid item>{showcase}</Grid>
                        </Zoom>
                    ))}
                </Grid>
            </Grid>
        </Grid>
    )
}
