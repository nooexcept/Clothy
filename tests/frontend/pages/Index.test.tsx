import React from 'react'
import { ShallowWrapper } from 'enzyme'
import { createShallow } from '@material-ui/core/test-utils'
import Index, { getServerSideProps } from '../../../pages/index'
import ShowcaseList from '../../../components/ShowcaseList'
import RecentViewed from '../../../components/RecentViewed'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { GetServerSidePropsContext } from 'next'
import { ParsedUrlQuery } from 'querystring'
import { IncomingMessage } from 'http'

const mockedProduct = [
    {
        name: 'Ultra chill jumpers',
        url: 'ultra-chill-jumpers',
        description: 'If you wanna look chill, you need ultra chill jumpers',
        stock: 6,
        price: 80,
        cod: 'JUMPERS002',
        images: ['/photo-of-woman-standing-on-staircase-3120339.jpg'],
    },
]

const mockedShowcased = [
    {
        name: 'Summer happiness',
        cover: 'covered-stylish-woman-in-sunglasses-looking-away-3819572.jpg',
        description: 'The summer happiness collection!',
        products: ['5edec7af4fd2c4068cc53644', '5edec8b84fd2c4068cc53645'],
    },
    {
        name: 'Brazil fashion',
        cover: 'photo-of-woman-wearing-white-shirt-3376116.jpg',
        description: 'The Brazil fashion collection!',
        products: ['5edec9354fd2c4068cc53646'],
    },
    {
        name: 'Vintage',
        cover: 'woman-wearing-a-hat-3482614.jpg',
        description: 'The vintage collection!',
        products: ['5edecad54fd2c4068cc53647'],
    },
]

let shallow: ReturnType<typeof createShallow>
let wrapper: ShallowWrapper
const mock = new MockAdapter(axios)

beforeEach(() => {
    shallow = createShallow()
    wrapper = shallow(<Index showcased={mockedShowcased} recentProducts={[]} />)
})

afterEach(() => {
    mock.reset()
})

describe('<Index /> rendering', () => {
    it('should render without problems', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should render a <ShowcaseList /> to show featured products', () => {
        expect(wrapper.find(ShowcaseList)).toHaveLength(1)
    })

    it('shouldn\'t render a <RecentViewed /> if it didn\'t receive any recent products', () => {
        expect(wrapper.find(RecentViewed)).toHaveLength(0)
    })

    it('should render a <RecentViewed /> if it did receive recent products', () => {
        const wrapperWithRecentProds = shallow(
            <Index
                showcased={mockedShowcased}
                recentProducts={[mockedProduct[0]]}
            />
        )
        expect(wrapperWithRecentProds.find(RecentViewed)).toHaveLength(1)
    })
})

type SessionMessage = Partial<IncomingMessage> & Partial<Express.Session>

describe('<Index /> data fetching', () => {
    it('should return only showcased if there is no recent product, in getServerSideProps', async () => {
        mock.onGet(
            `http://localhost:${process.env.EXPRESS_PORT}/api/collections/showcases`
        ).reply(200, mockedShowcased)

        const mockedSession: {
            req: { session: SessionMessage }
        } = {
            req: {
                session: {},
            },
        }

        const res = await getServerSideProps(
            (mockedSession as unknown) as GetServerSidePropsContext<
                ParsedUrlQuery
            >
        )

        expect(res.props.recentProducts).toEqual(null)
        expect(res.props.showcased.length).toEqual(mockedShowcased.length)

        for (let i = 0; i < mockedShowcased.length; i += 1)
            expect(res.props.showcased[i]).toEqual(mockedShowcased[i])
    })

    it('should return showcased and recent product if provided, in getServerSideProps', async () => {
        mock.onGet(
            `http://localhost:${process.env.EXPRESS_PORT}/api/collections/showcases`
        ).reply(200, mockedShowcased)

        mock.onGet(
            `http://localhost:${process.env.EXPRESS_PORT}/api/products/${mockedProduct[0].url}`
        ).reply(200, mockedProduct)

        const mockedSession: {
            req: { session: SessionMessage }
        } = {
            req: {
                session: {
                    recentProducts: [mockedProduct[0].url],
                },
            },
        }

        const res = await getServerSideProps(
            (mockedSession as unknown) as GetServerSidePropsContext<
                ParsedUrlQuery
            >
        )

        expect(res.props.recentProducts.length).toEqual(1)
        expect(res.props.recentProducts[0]).toEqual(mockedProduct)

        expect(res.props.showcased.length).toEqual(mockedShowcased.length)

        for (let i = 0; i < mockedShowcased.length; i += 1)
            expect(res.props.showcased[i]).toEqual(mockedShowcased[i])
    })
})
