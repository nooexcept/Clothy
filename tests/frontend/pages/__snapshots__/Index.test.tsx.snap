// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Index /> rendering should render without problems 1`] = `
<PublicLayout>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <ShowcaseList
        showcases={
          Array [
            <Showcase
              description="The summer happiness collection!"
              img="covered-stylish-woman-in-sunglasses-looking-away-3819572.jpg"
              title="Summer happiness"
            />,
            <Showcase
              description="The Brazil fashion collection!"
              img="photo-of-woman-wearing-white-shirt-3376116.jpg"
              title="Brazil fashion"
            />,
            <Showcase
              description="The vintage collection!"
              img="woman-wearing-a-hat-3482614.jpg"
              title="Vintage"
            />,
          ]
        }
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</PublicLayout>
`;
